Web: Use Intl.ListFormat rather than .join(", ")

CIC4DA opened the pull request:
<!-- Describe your pull request here.-->
This PR brings internationalization by replacing the inconsistent .join(", ") with Intl.ListFormat for accurate and localized list formatting, using a common utility function.

Screenshot of Visual Changes:
![visual_changes_direct_message](https://github.com/zulip/zulip/assets/91906953/5b711248-65dc-4e0e-b788-6b014bddcc4e)

Edit : Removed Internationalization from left side bar.
![image](https://github.com/zulip/zulip/assets/91906953/24ec45b0-c8e9-4b48-97e9-bce63a431be5)



- [x] Files which were to be internationalized are using Intl.ListFormat.
- [x]  Uses ```{ style: "long", type: "conjunction" }``` could be changed according to use.
- [x]  Add a fallback system for MacOs. 
- [x]  Refactor tests to test with the new changes.

Fixes: #26936 Use Intl.ListFormat rather than .join(", ")

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on web/src/util.ts:
```diff
@@ -489,3 +490,21 @@ export function is_valid_url(url: string, require_absolute = false): boolean {
+    // If Intl.ListFormat is not supported
+    if (!Intl) {
+        return array.join(", ");
+    }
```
Hmm, does this address the bug we saw in 2fc327b775ea4aadec6a47242cef8ef6fce01fce correctly?

I'm not sure if it's possible for `Intl` to exist without `Intl.ListFormat` existing.

CIC4DA replied:
You're correct; there are situations where Intl may exist without Intl.ListFormat. 
To address this, I've updated the check to verify that Intl.ListFormat is undefined instead of checking for the existence of Intl.

timabbott commented on web/src/filter.ts:
```diff
@@ -980,7 +980,7 @@ export class Filter {
             // and also to ensure that we return a string and not an array so that we
             // can have the same return type as other cases.
-            return names.join(", ");
+            return util.format_array_as_list(names, "long", "conjunction");
```
We can delete delete the comment detailing he previous logic.

CIC4DA replied:
Done.

timabbott commented on web/tests/realm_playground.test.js:
```diff
@@ -76,14 +76,14 @@ run_test("get_pygments_typeahead_list_for_settings", () => {
     assert.equal(iterator.next().value[1], $t({defaultMessage: "Custom language: custom_lang"}));
     assert.equal(iterator.next().value[1], $t({defaultMessage: "Custom language: invent_a_lang"}));
-    assert.equal(iterator.next().value[1], "JavaScript (javascript, js, javascript, js)");
+    assert.equal(iterator.next().value[1], "JavaScript (javascript, js, javascript, and js)");
```
Looking at these tests ... I think we are missing a deduplication+sorting step, as we shouldn't have `js` appear twice in a given list of what language ode generate something... or maybe we just have everything twice for some reason?

Want to add a bonus commit fixing that?

timabbott replied:
Let's fix this in a new commit.

CIC4DA replied:
Done.

timabbott commented on web/tests/pm_list_data.test.js:
```diff
@@ -233,18 +233,18 @@ test("get_list_info_unread_messages", ({override}) => {
-        "Alice, Bob",
+        "Alice and Bob",
         "Bob",
     ]);
```
Do the test changes in this commit go with the `inbox_ui.js` changes?

`pm_list_data` doesn't import that module....

CIC4DA replied:
My bad, I found that it should come with the commit of people.ts, as the change is caused due the change in get_recipients function of people.ts
I'll add it with commit for people.ts.

CIC4DA replied:
I found one more test ```buddy_data.test.ts``` which was in wrong commit. Updated it with ```People``` commit.

timabbott commented on web/tests/message_store.test.js:
```diff
@@ -132,7 +132,7 @@ test("process_new_message", () => {
     assert.equal(message.reply_to, "bob@example.com,cindy@example.com");
     assert.equal(message.to_user_ids, "103,104");
-    assert.equal(message.display_reply_to, "Bob, Cindy");
+    assert.equal(message.display_reply_to, "Bob and Cindy");
```
The `web/src/message_notifications.js` code path appears to use this field for non-display purposes to construct unique keys; I think that might now be incorrect

The `web/src/recent_view_ui.js` code also seems to sort by `message.display_reply_to`; does that seem correct to you? Might require some auditing.

CIC4DA replied:
In ```web/src/message_notifications.js```, we can convert back the internationalized list to an array and then to a list using .join(", "), and then set the key.

Something like this :

```
const reconstructed_list = message.display_reply_to
                          .replace('and', '') 
                          .split(',')
                          .map(item => item.trim())
                          .join(", ");

key = reconstructedList;
```

I'm not sure about ```web/src/recent_view_ui.js```, as it uses ```message.display_reply_to``` for sorting and ```message.display_reply_to``` are already sorted name, with an extra ```and``` in it. Will it effect the sorting of multiple strings? 
What is your view about it?

timabbott commented on web/e2e-tests/drafts.test.ts:
```diff
@@ -177,7 +177,7 @@ async function test_restore_private_message_draft_via_draft_overlay(page: Page):
     assert.strictEqual(
         await common.get_text_from_selector(page, "title"),
-        "Cordelia, Lear's daughter, King Hamlet - Zulip Dev - Zulip",
+        "Cordelia, Lear's daughter and King Hamlet - Zulip Dev - Zulip",
```
This hunk should presumably be squashed into whatever commit caused the result of this test to change.

CIC4DA replied:
It was caused due to ```web/src/people.ts```. Squashed with ```people.ts``` commit.

CIC4DA commented:
@timabbott, I have done the following:

- [ ] Addressed the feedback.
- [ ] Each test change is included in the commit of file which caused the change.
- [ ] Added a screenshot of the visual changes.
- [ ] Removed ```message_store.display_replay_to``` commit from this PR

timabbott commented on web/src/realm_playground.ts:
```diff
@@ -90,7 +91,10 @@ export function get_pygments_typeahead_list_for_settings(query: string): Map<str
+        language_labels.set(
+            key,
+            key + " (" + util.format_array_as_list(values, "long", "conjunction") + ")",
+        );
```
I think after some consultation with @alya, for this one we want to use the more compact format using just commas -- i.e. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat/ListFormat#short I think?

CIC4DA replied:
Done.

timabbott commented:
I extracted several commits to https://github.com/zulip/zulip/pull/28611 to merge those; posting comments on the rest.

timabbott commented on web/src/composebox_typeahead.js:
```diff
@@ -1235,6 +1236,6 @@ export function initialize({on_enter_send}) {
         const recipients = typeahead_helper.get_cleaned_pm_recipients(val);
-        $(this).val(recipients.join(", "));
+        $(this).val(util.format_array_as_list(recipients, "long", "conjunction"));
     });
```
Do you know what this block does? I wonder if it's stale code predating the input pill code; @evykassirer FYI in case you know. 

(In any case, my actual question is what this looks like, so a screenshot would be great)

evykassirer replied:
(I'm not very familiar with with the private message pill code yet, so I don't think I can help here without diving further into it)

CIC4DA replied:
According to me, it is a stale code which is predating the input pill code. As we can see it's effect has to be shown here: 
![image](https://github.com/zulip/zulip/assets/91906953/4e81516e-a77f-44e5-afa3-1bd50cd8e86c)

But it is overridden by input pill code.

The usage of it can be seen in ```compose.hbs``` 

``` html 
<div id="compose-direct-recipient" class="pill-container" data-before="{{t 'You and' }}">
                                <div class="input" contenteditable="true" id="private_message_recipient" data-no-recipients-text="{{t 'Add one or more users' }}" data-some-recipients-text="{{t 'Add another user...' }}"></div>
                            </div>

timabbott replied:
Yeah so probably the right answer is to delete that function instead of updating it, in its own commit with a nice commit message.

CIC4DA replied:
Done.

timabbott commented on web/src/input_pill.ts:
```diff
@@ -268,7 +269,7 @@ export function create<T>(opts: InputPillCreateOptions<T>): InputPillContainer<T
             );
 
-            store.$input.text(drafts.join(", "));
+            store.$input.text(util.format_array_as_list(drafts, "long", "conjunction"));
```
Do we have an example of this code path in action? It might make sense to use the short format here, since pills are small, or it might be the case that this is supposed to be a `,`, if it's meant to be parsed by software.

CIC4DA replied:
I think it should be using ```.join(",")``` as it's meant to be parsed by software.

timabbott replied:
OK let's do that then.

CIC4DA replied:
Done.

timabbott commented on web/tests/pm_list_data.test.js:
```diff
@@ -111,7 +111,7 @@ test("get_conversations", ({override}) => {
         },
         {
-            recipients: "Alice, Bob",
+            recipients: "Alice and Bob",
```
So I think we probably want the `short` / comma-separated format for the left sidebar at the very least, since there's not very much space there and we'd rather use it for actual content.

This is probably a case where we will want to revisit the decision after completing projects like #27146 that make space less of a constraint.

CIC4DA replied:
If we want to just stick to short / comma-separated format in left sidebar I have to refactor ```people.ts``` ```get_recipients``` function. And make it return an array of names, and then join them using ```util.format_array_as_list``` to the desired format.

Do we want to use ```short``` format: 
![image](https://github.com/zulip/zulip/assets/91906953/5b6884e0-f5a3-4bd7-bdec-1d853fce557d)

Or use "narrow" format: 
![image](https://github.com/zulip/zulip/assets/91906953/2ea91ad3-132e-415f-9a6b-45a8c70c0b7e)

CIC4DA replied:
As of now I have used ```Short``` format here.

timabbott replied:
@alya thoughts on which format you prefer?

alya replied:
Ah, sorry I missed the question earlier. `narrow` seems fine. We don't use `&` elsewhere, and I don't think it's an improvement.

CIC4DA replied:
Done.

CIC4DA commented:
@timabbott I have made some required changes, and requires suggestion from you on above reviews, kindly take a look. Thanks.

CIC4DA commented on web/tests/typeahead_helper.test.js:
The function ```get_cleaned_pm_recipients``` of ```typehead_helper.ts```, was only used in .on(blur) function. But I didn't removed it as it can be used further in future. And to cover it in test, Added a test for ```get_cleaned_pm_recipients``` in ```typehead_helper.test.js```

whilstsomebody commented:
Hey @CIC4DA , can I continue with your work ?

CIC4DA commented:
Yea, no problem

timabbott commented:
Closing in favor of #32872, which completes this. Thanks for the effort @CIC4DA!
